core/model/src/main/java/org/eclipse/rdf4j/model/impl/NumericLiteral.java

```diff
@@ -7,32 +7,41 @@
  *******************************************************************************/
 package org.eclipse.rdf4j.model.impl;
 
+import java.util.Objects;
+import java.util.Optional;
+
 import org.eclipse.rdf4j.model.IRI;
+import org.eclipse.rdf4j.model.base.AbstractLiteral;
 import org.eclipse.rdf4j.model.datatypes.XMLDatatypeUtil;
 import org.eclipse.rdf4j.model.vocabulary.XSD;
 
 /**
- * An extension of {@link SimpleLiteral} that stores a numeric value to avoid parsing.
+ * An extension of {@link AbstractLiteral} that stores a numeric value to avoid parsing.
  *
  * @author David Huynh
+ * @author Jerven Bolleman
  */
-public class NumericLiteral extends SimpleLiteral {
+public class NumericLiteral extends AbstractLiteral {
 
 	private static final long serialVersionUID = 3004497457768807919L;
 
 	private final Number number;
 
+	private final IRI datatype;
+
 	/**
 	 * Creates a literal with the specified value and datatype.
 	 */
 	protected NumericLiteral(Number number, IRI datatype) {
-		super(XMLDatatypeUtil.toString(number), datatype);
+		super();
+		assert Objects.nonNull(number);
+		assert Objects.nonNull(datatype);
+		this.datatype = datatype;
 		this.number = number;
 	}
 
 	protected NumericLiteral(Number number, XSD.Datatype datatype) {
-		super(XMLDatatypeUtil.toString(number), datatype);
-		this.number = number;
+		this(number, datatype.getIri());
```

Comment:
This nullifies the sorting performance improvements that depend on the XSD.Datatype enum. If we end up merging https://github.com/eclipse/rdf4j/pull/3596 then it won't matter, but if not it would be worthwhile to implement support for XSD.Datatype here in this class.

Reply:
I would like to merge the XSD.Datatype work into this when it is ready.